lib/puma/stream_client.rb

```diff
@@ -0,0 +1,69 @@
+module Puma
+  # This serves as a base class for any plugin that wants to take over the
+  # socket and wait on IO.
+  #
+  # The subclasses are expected to implement:
+  #
+  #  1. Methods to respond to changes in the underlying socket. These are
+  #     `#on_read_ready`, `#on_broken_pipe` & `#on_shutdown`.
+  #  2. A `#churn` method that runs within the thread pool, this is what can
+  #     be used to invoke app's logic.
+  #
+  # The underlying socket is available through `@io`.
+  #
+  # The other methods should never be overriden.
+  class StreamClient
+    def initialize(io)
+      @io = io
+    end
+
+    def to_io
+      @io
+    end
+
+    def timeout_at
+      false
+    end
+
+    def closed?
+      @io.closed?
+    end
+
+    def stream?
+      true
+    end
+
+    # This method will be invoked when the IO descriptor has new data pending
+    # to be read. You can read from `@io` at this time.
+    #
+    # You can return a truthy value to add this client to the thread pool.
+    def on_read_ready
+      raise NotImplementedError
+    end
+
+    # This method will be invoked when the underlying connection is broken
+    # for whatever reason (client timeout, abrupt client disconnection, etc.).
+    #
+    # You can return a truthy value to add this client to the thread pool.
+    def on_broken_pipe
+      raise NotImplementedError
+    end
+
+    # This method will be invoked when the server is being stopped.
+    #
+    # It's not possible to run more work on the thread pool at this stage.
+    def on_shutdown
+      raise NotImplementedError
+    end
+
+    # This is the method that the thread pool will be consuming. A "churn" is
+    # enqueued on the thread pool each time `#churn` or `#read_more` return a
+    # truthy value.
+    #
+    # You can return a truthy value to add this client to the thread pool
+    # again.
+    def churn
```

Comment:
I don't understand how `churn` is used. Can you give an example?

Reply:
This'd be the equivalent of invoking the Rack application in traditional HTTP requests. I.e. Puma calls `churn` on the plugin as it calls `call` on the web application.

When `churn` on the plugin is called, the plugin can run the web application. This is executed inside Puma's thread pool.

![puma-plugin-churn](https://user-images.githubusercontent.com/23031/62141592-d47af580-b2ba-11e9-9e3e-65181c7d1eb5.png)

An example of how this is used: https://github.com/Jesus/puma-websockets/blob/22985041bd41e4488231c34d73af4730dfa46d58/lib/puma/websockets/web_socket_client.rb#L84-L89

A possible follow up question would be: Why isn't `on_read_ready` called directly inside the thread pool? Because the plugin may decide that there's no need to invoke the web application for example if what we've received is just a Ping message in a WebSocket.

Reply:
Nice diagram!

Reply:
Thanks! I think something like that should be included in the final PR before merging.